File: ImperatorToCK3/Source/Imperator/Genes/AccessoryGeneTemplate.cpp
#include "AccessoryGeneTemplate.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"
#include "WeightBlock.h"


Imperator::AccessoryGeneTemplate::AccessoryGeneTemplate(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}

void Imperator::AccessoryGeneTemplate::registerKeys()
{
	registerKeyword("index", [this](std::istream& theStream) {
		index = commonItems::getInt(theStream);
	});
	registerRegex("male|female|boy|girl", [this](const std::string& ageSexStr, std::istream& theStream) {
		const auto sexAgeStr = commonItems::stringOfItem(theStream).getString();
		std::stringstream tempStream(sexAgeStr);
		if (sexAgeStr.find('{') != std::string::npos) // for full blocks: "male = { 6 = hoodie 7 = tshirt }"
		{
			auto ageSexBlock = std::make_shared<WeightBlock>(tempStream);
			ageSexWeightBlocks.insert(std::pair(ageSexStr, ageSexBlock));
		}
		else // for copies: "boy = male"
		{
			const auto sexAge = commonItems::getString(tempStream);
			if (ageSexWeightBlocks.contains(sexAge))
				ageSexWeightBlocks.insert(std::pair(ageSexStr, ageSexWeightBlocks.find(sexAge)->second));
		}
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Genes/AccessoryGeneTemplate.cs
using commonItems;
using System.Collections.Generic;

namespace ImperatorToCK3.Imperator.Genes {
	public class AccessoryGeneTemplate : Parser {
		public uint Index { get; private set; } = 0;
		public Dictionary<string, WeightBlock> AgeSexWeightBlocks { get; private set; } = new();

		public AccessoryGeneTemplate(BufferedReader reader) {
			RegisterKeys();
			ParseStream(reader);
			ClearRegisteredRules();
		}
		private void RegisterKeys() {
			RegisterKeyword("index", reader => Index = (uint)reader.GetInt());
			RegisterRegex("male|female|boy|girl", (reader, ageSexStr) => {
				var stringOfItem = new StringOfItem(reader).ToString();
				var tempStream = new BufferedReader(stringOfItem);
				if (stringOfItem.Contains('{')) { // for full blocks: "male = { 6 = hoodie 7 = tshirt }"
					var ageSexBlock = new WeightBlock(tempStream);
					AgeSexWeightBlocks.Add(ageSexStr, ageSexBlock);
				} else { // for copies: "boy = male"
					var sexAge = tempStream.GetString();
					if (AgeSexWeightBlocks.TryGetValue(sexAge, out var blockToCopy)) {
						AgeSexWeightBlocks.Add(ageSexStr, blockToCopy);
					}
				}
			});
			RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
		}
	}
}
